tests/utilities/fault_tolerant.py

```diff
@@ -0,0 +1,142 @@
+            log.info("READY TO BE KILLED WITH SIGTERM SIGNAL.")
+            while not self.trainer._terminate_gracefully:
+                sleep(0.1)
+        batch = batch["data"] if isinstance(batch, dict) else batch
+        self.seen_batches.append(torch.stack(batch) if isinstance(batch, list) else batch)
+        loss = sum(self.layer(b).sum() for b in batch)
+        return loss
+
+    def configure_optimizers(self):
+        return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+
+class RandomGetItemDataset(Dataset):
+    """A dataset with random elements generated using global rng from torch, numpy and python."""
+
+    def __init__(self, length, size):
+        self.size = size
+        self.len = length
+
+    def __getitem__(self, index):
+        t = torch.rand(self.size)
+        n = torch.from_numpy(np.random.rand(self.size))
+        p = torch.tensor([python_random.random() for _ in range(self.size)])
+        sample = (index + (t + n + p) / 10).float()
+        return sample
+
+    def __len__(self):
+        return self.len
+
+
+def _run_training(trainer_kwargs, dataset_classes, fail_on_step: int = -1, ckpt_path=None):
+    seed_everything(1)
+    train_dataloader = [
+        DataLoader(dataset_class(3, 1), batch_size=1, num_workers=0) for dataset_class in dataset_classes
+    ]
+    train_dataloader = train_dataloader[0] if len(train_dataloader) == 1 else train_dataloader
+    model = Model(fail_on_step=fail_on_step)
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model, train_dataloaders=train_dataloader, ckpt_path=ckpt_path)
+    return model.seen_batches, model.parameters()
+
+
+def main():
+
+    tmpdir = "."
+
+    os.makedirs(tmpdir, exist_ok=True)
+
+    env_backup = os.environ.copy()
+
+    auto_restart_checkpoint_path = os.path.join(tmpdir, ".pl_auto_save.ckpt")
+    auto_restart_checkpoint_path_exists = os.path.exists(auto_restart_checkpoint_path)
+
+    seed_everything(42)
+
+    os.environ["PL_FAULT_TOLERANT_TRAINING"] = "1"
+
+    dataset_classes = [RandomGetItemDataset]
+
+    trainer_kwargs = dict(
+        default_root_dir=tmpdir,
+        max_epochs=3,
+        enable_progress_bar=False,
+        enable_model_summary=False,
+    )
+
+    if auto_restart_checkpoint_path_exists:
+        fail_on_step = -1
+        completed_batches = 4
+    else:
+        fail_on_step = 4
+        completed_batches = 5
+
+    # Perform a failure
+    complete_batches, weights = _run_training(trainer_kwargs, dataset_classes, fail_on_step=fail_on_step)
+    assert len(complete_batches) == completed_batches
+
+    if not auto_restart_checkpoint_path_exists:
+        checkpoint_path = os.path.join(tmpdir, ".pl_auto_save.ckpt")
+        assert os.path.exists(checkpoint_path)
+        log.info(".pl_auto_save.ckpt exists.")
+
+    if auto_restart_checkpoint_path_exists:
+        params = [w for w in weights]
+        log.info(params)
+        assert "[-1.0939, -0.4306]" in str(params[-1])
```

Comment:
This could change with different torch versions.

Reply:
Removed !